proposals.go

```diff
@@ -0,0 +1,165 @@
+	"net/http"
+	"net/url"
+)
+
+const (
+	apiEndpoint = "proposals.decred.org"
+)
+
+type Politea struct {
+	csrfToken    string
+	serverPolicy *ServerPolicy
+}
+
+func newPolitea() Politea {
+	return Politea{}
+}
+
+func (p *Politea) prepareRequest(path, method string, body []byte) (*http.Request, error) {
+	req := &http.Request{
+		Method: method,
+		URL:    &url.URL{Scheme: "https", Host: "proposals.decred.org", Path: "/api/v1" + path},
+	}
+
+	if body != nil {
+		req.Body = ioutil.NopCloser(bytes.NewBuffer(body))
+	}
+
+	if method == "POST" {
+		originalURL := req.URL
+		if p.csrfToken == "" {
+			req.URL.Path = "/api/v1/version"
+			res, err := http.DefaultClient.Do(req)
+			if err != nil {
+				return nil, fmt.Errorf("error fetching csrf token")
+			}
+
+			err = p.handleResponse(res, &p.csrfToken)
+			if err != nil {
+				return nil, err
+			}
+		}
+		req.URL = originalURL
+		req.Header.Set("X-CSRF-TOKEN", p.csrfToken)
+	}
+
+	return req, nil
+}
+
+func (p *Politea) makeRequest(path, method string, body []byte, dest interface{}) error {
+	req, err := p.prepareRequest(path, method, body)
+	if err != nil {
+		return err
+	}
+
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return err
+	}
+
+	return p.handleResponse(res, dest)
+}
+
+func (p *Politea) handleResponse(res *http.Response, dest interface{}) error {
+	if res.StatusCode == 200 {
+		return json.NewDecoder(res.Body).Decode(dest)
+	}
+
+	body, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return fmt.Errorf("error reading response body: %s", err.Error())
+	}
+	res.Body.Close()
+
+	return fmt.Errorf("request error: %s", string(body))
+}
+
+func (p *Politea) getServerPolicy() (*ServerPolicy, error) {
+	var serverPolicy ServerPolicy
+
+	err := p.makeRequest("/policy", "GET", nil, &serverPolicy)
+	if err != nil {
+		return nil, fmt.Errorf("error fetching politea policy: %v", err)
+	}
+
+	return &serverPolicy, nil
+}
+
+func (p *Politea) getProposalsChunk(startHash string) ([]Proposal, error) {
+	var result Proposals
+
+	path := "/proposals/vetted"
+	if startHash != "" {
+		path = fmt.Sprintf(path+"?after=%s", startHash)
+	}
+
+	err := p.makeRequest(path, "GET", nil, &result)
+	if err != nil {
+		return nil, fmt.Errorf("error fetching proposals from %s: %v", startHash, err)
+	}
+
+	return result.Proposals, err
+}
+
+// GetProposalsChunk gets proposals starting after the proposal with the specified
+// censorship hash. The number of proposals returned is specified in the poltiea
+// policy API endpoint
+func (p *Politea) GetProposalsChunk(startHash string) (string, error) {
+	proposals, err := p.getProposalsChunk(startHash)
+	if err != nil {
+		return "", err
+	}
+
+	jsonBytes, err := json.Marshal(proposals)
+	if err != nil {
+		return "", fmt.Errorf("error marshalling proposal result to json: %s", err.Error())
+	}
+
+	return string(jsonBytes), nil
+}
+
+// GetAllProposal fetches all vetted proposals from the API
+func (p *Politea) GetAllProposals() (string, error) {
```

Comment:
This only fetches 20 proposals which I believe is limited by the server policy, it should fetch all proposals

Reply:
This should also follow a similar pattern as fetching transactions, where the offset, limit, filters and newest first flag can be set https://github.com/planetdecred/dcrlibwallet/blob/master/transactions.go#L96